### PR TITLE
Improve checking of complex rest parameter types

### LINKED ISSUE #26110: Spreads instantiated with array subtypes types exhibit bad behavior
Found by playing around with @mattmccutchen's example at https://github.com/Microsoft/TypeScript/issues/26013#issuecomment-409330936:

```ts
interface CoolArray<E> extends Array<E> { 
    hello: number;
}

declare function foo<T extends any[]>(cb: (...args: T) => void): void;

foo<CoolArray<any>>();    // error
foo<CoolArray<any>>(100); // error
foo<CoolArray<any>>(foo); // no error!!
```

**Expected**: All of these calls produce an error and have a good error message.
**Actual**: The first two have bad errors, the last one is considered okay.

### LINKED ISSUE #26491: Generic parameters constrained to unions of tuples are not inferred correctly in rest parameter positions
**TypeScript Version:**  3.1.0-dev.20180813

**Search Terms:** 
Tuple rest generic union

**Code**
```ts
declare function hmm<A extends [] | [number, string]>(...args: A): void;
hmm(); // okay, A = []
hmm(1, "s"); // okay, A = [1, "s"]
hmm("what"); // no error?  A = [] | [number, string] ?
```

**Expected behavior:**
Error on `hmm("what")`.  `A` should be inferred as `[string]`, which does not extend `[] | [number, string]`.

**Actual behavior:**
No error on `hmm("what")`.  Apparently `A` is inferred as the constraint type `[] | [number, string]` despite that not matching the actual argument type.

**Playground Link:** 
[Link](https://www.typescriptlang.org/play//#src=declare%20function%20hmm%3CA%20extends%20%5B%5D%20%7C%20%5Bnumber%2C%20string%5D%3E%28...args%3A%20A%29%3A%20void%3B%0A%0Ahmm%28%29%3B%20%2F%2F%20okay%2C%20A%20%3D%20%5B%5D%0Ahmm%281%2C%20%22s%22%29%3B%20%2F%2F%20okay%2C%20A%20%3D%20%5B1%2C%20%22s%22%5D%0Ahmm%28%22what%22%29%3B%20%2F%2F%20no%20error%3F%20%20A%20%3D%20%5B%5D%20%7C%20%5Bnumber%2C%20string%5D%20%3F)

**Related Issues:**
#24897: introduction of tuples into rest/spread expressions
#26110: similar issue with bad or weird inference behavior on rest arguments

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we properly check rest parameters of any subtype of `any[]`, including user-defined array types and unions of tuple types. In cases where a rest parameter type is not exactly a `T[]` or a tuple type, we now collect the corresponding remainder of the argument list as a tuple type and then check that this tuple type is assignable to the rest parameter type.

```ts
declare const foo: (x: string, ...args: [string] | [number, boolean]) => void;
declare const t: [string] | [number, boolean];

foo("abc", "def");  // Ok
foo("abc", 10, true);  // Ok
foo("abc", ...t);  // Ok
foo("abc", 10);  // Error, [number] not assignable to [string] | [number, boolean]
foo("abc");  // Error, [] not assignable to [string] | [number, boolean]
```

Effectively, rest parameters typed as unions of tuple types provide a form of overloading expressed in a single function type signature. Unlike regular overloading, generic rest parameters can be used to form composed overloads:

```ts
type Func<T extends any[]> = (x: string, ...args: T) => void;

declare const f: Func<[] | [string] | [number, ...boolean[]]>;

f("abc");
f("abc", "def");
f("abc", 10);
f("abc", 10, true);
f("abc", 10, true, false);
f();  // Error
f("abc", true);  // Error
```

Function type relations have been improved to handle rest parameters of union types:

```ts
declare let f1: (x: string, ...args: [string] | [number, boolean]) => void;
declare let f2: (x: string, y: string) => void;
declare let f3: (x: string, y: number, z: boolean) => void;
declare let f4: (...args: [string, string] | [string, number, boolean]) => void;

f2 = f1;  // Ok
f3 = f1;  // Ok
f4 = f1;  // Error, misaligned complex rest types
f1 = f2;  // Error
f1 = f3;  // Error
f1 = f4;  // Error, misaligned complex rest types
```

Note that we currently don't "normalize" misaligned rest parameters. It would certainly be possible to do so, but it is not clear that the additional complexity is merited.

Fixes #26110.
Fixes #26491.

### CODE REVIEW

**Comment by mattmccutchen** [comment]
I just tested the PR (commit 349bee9) on the bad example from https://github.com/Microsoft/TypeScript/issues/26110#issuecomment-415199148 and I get no compile errors.  Is this expected?

**Comment by mattmccutchen** [comment]
For anyone watching, the discussion of the unsoundness continues at https://github.com/Microsoft/TypeScript/issues/26110#issuecomment-416076790.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -12693,6 +12699,11 @@ namespace ts {
             return type.target.hasRestElement ? type.typeArguments![type.target.typeParameters!.length - 1] : undefined;
         }
 
+        function getRestArrayTypeOfTupleType(type: TupleTypeReference) {
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/checker.ts]: I like it where it is, it is closely related to the function just above.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -10718,9 +10722,10 @@ namespace ts {
             }
 
             const sourceCount = getParameterCount(source);
-            const sourceGenericRestType = getGenericRestType(source);
-            const targetGenericRestType = sourceGenericRestType ? getGenericRestType(target) : undefined;
-            if (sourceGenericRestType && !(targetGenericRestType && sourceCount === targetCount)) {
+            const sourceRestType = getNonArrayRestType(source);
+            const targetRestType = getNonArrayRestType(target);
+            if (sourceRestType && targetRestType && sourceCount !== targetCount) {
+                // We're not able to relate misaliged complex rest parameters
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/checker.ts]: Will fix.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -18814,28 +18834,27 @@ namespace ts {
...
+                    // parameter types yet and therefore excess property checks may yield false positives (see #17041).
+                    const checkArgType = excludeArgument ? getRegularTypeOfObjectLiteral(argType) : argType;
+                    if (!checkTypeRelatedTo(checkArgType, paramType, relation, reportErrors ? arg : undefined, headMessage)) {
+                        return false;
                     }
                 }
             }
+            if (restType) {
+                const spreadType = getSpreadArgumentType(args, argCount, args.length, restType, /*context*/ undefined);
+                const errorNode = reportErrors ? argCount < args.length ? args[argCount] : node : undefined;
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/checker.ts]: Sort of agree, but the rewrite you suggest doesn't work because it injects a `false` into the union type. Ima keepit.

**Comment by CurtisDS** [comment]
What version of Typescript is this fix implemented in. Im running version 3.1.1 and still get and error when I try this

```
type Foo = [number, number];
type Bar = number;
type FooFoo = [Foo, Foo];
type BarBar = [Bar, Bar];
type FooFooBarBar = FooFoo | BarBar;

function fn(...args: FooFoo | BarBar): void { // error: "A rest parameter must be of an array type."
}
```

**Pull request merged**